js/general_patterns.md

```diff
@@ -0,0 +1,210 @@
+something as trivial as a `write()` operation:
+
+```js
+net.connect(8000, function(err) {
+  let writeReq = this.write('blah');
+});
+```
+
+Each handle and request already require these objects be created. So it is no
+performance disadvantage to return these to the user. While it does offer
+greater flexibility in both the amount of information we can communicate and
+how it can be used.
+
+For example, as the API evolves a method may be added that returns the number
+of bytes processed on the individual request. When data is written the kernel
+may only be able to receive so much. Allowing to check how much of the data has
+been written out to the kernel. This could be used to inform on the status of a
+large upload.
+
+Requests also have their own events. It should be possible to track errors from
+individual requests. Along with being notified when the data has been written:
+```js
+net.connect(8000, function(err) {
+  let writeReq = this.write('blah');
+
+  writeReq.onerror(onWriteError);
+  writeReq.oncomplete(onWriteComplete);
+});
+```
+
+Now, why not just use the error-back method in this case? Because it's very
+likely developers only want to know when something has gone wrong. Forcing the
+callback to be called for every request could be a significant strain on
+resources when there is a high volume of requests.
+
+Since the object context always represents the `this` of a request, I have
+considered a standard object property that can be accessed to represent the
+handle that was used to send the request. It could be gotten around by simply
+passing the handle as an argument, but since it's likely operations on the
+handle will be necessary from the callback making it available by default in
+some form seems appropriate.
+
+
+
+### Exception Handling
+
+Exceptions are only thrown when absolutely necessary. In the case of
+asynchronous calls, except when critical, the error is passed to the callback.
+Though if no callback is passed to the asynchronous call then throwing is
+inevitable. Basically, exceptions should only be thrown when there is no other
+possible execution path to take.
+
+
+
+### Type Coercion
+
+The only coercion allowed is from string to number. This is always done using
+the `+` operator. For example:
+```
+function caller(n) {
+  if (typeof n === 'string')
+    n = +n;
+}
+```
+
+Reason for using the `+` operator is because it supports operations that both
+`parseInt()` and `parseFloat()` don't support simultaneously.
+
+- `parseFloat('0xff') === 0`
+
+- `parseInt('0.1') === 0`
+
+The `+` operator handles both these cases.
+
+Testing will be easier with more explicit argument values.
+
+
+
+### Event Emitter
+
+There will be no actual event emitter library. The pattern is as follows:
+```
+blah.onready(callback[, ...args]);
+```
+
+Initially an event can only reference one callback. This is subject to change
+in the future, but for the time being will be easier to implement and test.
```

Comment:
Oh, the dots were just for some context without copying too much text. The change is really only to add the word `it` before `will`